[DOC] Update segment replication to include alerting plugin limitation (#4976)
opensearch-project/documentation-website: _tuning-your-cluster/availability-and-recovery/segment-replication/index.md

# Segment replication

> Segment replication involves copying segment files across shards instead of indexing documents on each shard copy. This approach enhances indexing throughput and reduces resource utilization, but increases network utilization. Segment replication is the first feature in a series of features designed to decouple reads and writes in order to lower compute costs.

Reviewer: ```suggestion
Segment replication involves copying segment files across shards instead of indexing documents on each shard copy. This approach enhances indexing throughput and reduces resource utilization but increases network utilization. Segment replication is the first feature in a series of features designed to decouple reads and writes in order to lower compute costs.
```

When the primary shard sends a checkpoint to replica shards on a refresh, a new segment replication event is triggered on replica shards. This happens:

- When a new replica shard is added to a cluster.
- When there are segment file changes on a primary shard refresh.
- During peer recovery, such as replica shard recovery and shard relocation (explicit allocation using the `move` allocation command or automatic shard rebalancing).

## Use cases

Segment replication can be applied in a variety of scenarios, including:

> - Users who have high write loads but do not have high search requirements and are comfortable with longer refresh times.

Reviewer: "Users" are not "scenarios", so the first two items do not complete the stem. The stem is fine, but I would rephrase the first two items.

Author: Revised.

Segment replication is the first feature in a series of features designed to decouple reads and writes in order to lower compute costs.

> **Limitation:** Segment replication is not compatible with the [Alerting]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) plugin. The Alerting plugin relies on the latest data being available on replica shards when using the `immediate` refresh policy. Segment replication does not provide the same guarantee, which means shards can become stale. This especially impacts [document-level monitors]({{site.url}}{{site.baseurl}}/observing-your-data/alerting/api/#document-level-monitors).   

Reviewer 2: >Segment replication does not provide the same guarantee, which means shards can become stale.

It is not clear what do you mean by same guarantee. Can you elaborate more on this ?

Reviewer 2: @lezzago @anasalkouz Can you also review this content.

Author: @dreamer-89 Revised. Please review for accuracy.

Author: Revised

> **Limitation:** Segment replication is not compatible with the [Alerting]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) plugin. The Alerting plugin relies on the latest data being available on replica shards when using the `immediate` refresh policy. The `immediate` refresh policy in segment replication can resource-intensive and introduce latency due to real-time data synchronization. As such, segment replication cannot guarantee the availability of the `immediate` refresh policy and thus lead to stale replica shards. This especially impacts [document-level monitors]({{site.url}}{{site.baseurl}}/observing-your-data/alerting/api/#document-level-monitors). Additionally, segment replication does not support the `wait_for` value in the `refresh` query parameter.   

Reviewer 3: Segment replication is not compatible with the [Alerting] is very broad term sentence. I think we need to be more specific. For example Segment replication is not compatible with the document-level monitors in Alerting Plugin.

Can you add that this is a temporary limitation? and we are going to fix this limitation in future releases by routing queries to primary shard

Author: Revised

> **Limitation:** Segment replication is not compatible with the [Alerting]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) plugin. The Alerting plugin relies on the latest data being available on replica shards when using the `immediate` refresh policy. The `immediate` refresh policy in segment replication can resource-intensive and introduce latency due to real-time data synchronization. As such, segment replication cannot guarantee the availability of the `immediate` refresh policy and thus lead to stale replica shards. This especially impacts [document-level monitors]({{site.url}}{{site.baseurl}}/observing-your-data/alerting/api/#document-level-monitors). Additionally, segment replication does not support the `wait_for` value in the `refresh` query parameter.   

Reviewer 4: We also need to call out that Security Analytics is not compatible at all with Segment replication since it heavily uses Document Level monitors behind the scenes.

Author: Revised

In segment replication, the primary shard is usually generating more network traffic than the replicas because it copies segment files to the replicas. Thus, it's beneficial to distribute primary shards equally between the nodes. To ensure balanced primary shard distribution, set the dynamic `cluster.routing.allocation.balance.prefer_primary` setting to `true`. For more information, see [Cluster settings]({{site.url}}{{site.baseurl}}/api-reference/cluster-api/cluster-settings/).

> Segment replication currently does not support the `wait_for` value in the `refresh` query parameter.

Reviewer 2: Do we need to keep all limitations at one place under limitation sections ?

Author: Yes, good catch. Change made.

Author: Do we have documentation for the refresh parameter?

Author: Revised

```
{% include copy.html %}

> This cluster-level setting cannot be enabled through the [REST API]({{site.url}}{{site.baseurl}}/api-reference/index/). This setting is not applied to system indexes and hidden indexes. By default, all system and hidden indexes in OpenSearch use document replication even if this setting is enabled.

Reviewer: ```suggestion
This cluster-level setting cannot be enabled through the [REST API]({{site.url}}{{site.baseurl}}/api-reference/index/). This setting is not applied to system indexes and hidden indexes. By default, all system and hidden indexes in OpenSearch use document replication, even if this setting is enabled.
```

1. Enabling segment replication for an existing index requires [reindexing](https://github.com/opensearch-project/OpenSearch/issues/3685).
1. [Cross-cluster replication](https://github.com/opensearch-project/OpenSearch/issues/4090) does not currently use segment replication to copy between clusters.
1. Segment replication is not compatible with [document-level monitors]({{site.url}}{{site.baseurl}}/observing-your-data/alerting/api/#document-level-monitors), which are used with the [Alerting]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) plugin. The Alerting plugin uses the latest available data on replica shards when using the `immediate` refresh policy, and segment replication can delay the policy's availability, resulting in stale replica shards.

> 1. Segment replication is not compatible with [security analytics]({{site.url}}{{site.baseurl}}/security-analytics/index/). Security analytics uses document-level monitors, which are not compatible with segment replication.  

Reviewer 3: Is this duplicate?

Author: Revised to combine into one bullet point.

> 1. Segment replication is not compatible with [security analytics]({{site.url}}{{site.baseurl}}/security-analytics/index/). Security analytics uses document-level monitors, which are not compatible with segment replication.  

Reviewer: Should both instances of "Security Analytics" be capitalized as a proper noun?

Author: Yes, as we're referring to the Security Analytics plugin.

1. Integration with remote-backed storage as the source of replication is [currently not supported](https://github.com/opensearch-project/OpenSearch/issues/4448). 

> 1. Read-after-write guarantees: The `wait_until` and `wait_for` refresh policy is not compatible with segment replication. If you use the `wait_until` refresh policy while ingesting documents, you'll get a response only after the primary node has refreshed and made those documents searchable. Replica shards will respond only after having written to their local translog. We are exploring other mechanisms for providing read-after-write guarantees. For more information, see the corresponding [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6046).  

Reviewer 5: Actually, the policy is internally has an enum value `WAIT_UNTIL` but the value set in a request is `wait_for`. So they are the same thing: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/support/WriteRequest.java#L97
```suggestion
1. Read-after-write guarantees: Segment replication currently does not support setting refresh policy to `wait_for`.  If you set the `refresh` query parameter to `wait_for` and then ingest documents, you'll get a response only after the primary node has refreshed and made those documents searchable. Replica shards will respond only after having written to their local translog. We are exploring other mechanisms for providing read-after-write guarantees. For more information, see the corresponding [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6046).  
```

> 1. Read-after-write guarantees: Segment replication currently does not support setting refresh policy to `wait_for`.  If you set the `refresh` query parameter to `wait_for` and then ingest documents, you'll get a response only after the primary node has refreshed and made those documents searchable. Replica shards will respond only after having written to their local translog. We are exploring other mechanisms for providing read-after-write guarantees. For more information, see the corresponding [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6046).  

Reviewer: ```suggestion
1. Read-after-write guarantees: Segment replication does not currently support setting the refresh policy to `wait_for`.  If you set the `refresh` query parameter to `wait_for` and then ingest documents, you'll get a response only after the primary node has refreshed and made those documents searchable. Replica shards will respond only after having written to their local translog. We are exploring other mechanisms for providing read-after-write guarantees. For more information, see the corresponding [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/6046).  
```

> 1. Segment replication is not compatible with [security analytics]({{site.url}}{{site.baseurl}}/security-analytics/index/). Security analytics relies on being able to access the latest data in an index, and segment replication can delay the availability of this data. 

Reviewer 5: Can we move this point to right after document-level monitors and say 
Segment replication is not compatible with security analytics because security analytics uses document-level monitors?

## Next steps

1. Track [future enhancements to segment replication](https://github.com/orgs/opensearch-project/projects/99).

> 1. Read this [blog post about segment replication](https://opensearch.org/blog/segment-replication/).

Reviewer: ```suggestion
1. Read [this blog post about segment replication](https://opensearch.org/blog/segment-replication/).
```
